### PR TITLE
Enforce type checks during tests

### PR DESCRIPTION
**Pull request by ymost**
I added enforcing of type checks during tests.
This uncovered a lot of bad type hints and a few bugs - all fixed now.

### CODE REVIEW

File: test/test_base.py [hunk 1]
```diff
@@ -29,7 +29,7 @@
 # Naming this variable "test_table" triggers the following bug: https://github.com/pytest-dev/pytest/issues/7378
 # noinspection PyTypeChecker
 mock_table = Table(
-    None, "mock_table",
+    Database(PyKustoClient('mock_cluster', fetch_by_default=False), 'mock_database', fetch_by_default=False), "mock_table",
```
Comment (ofrikleinfeld) [inline test/test_base.py:32]: is this change related to the typing fixes?

**Review by ofrikleinfeld — CHANGES_REQUESTED** [review]
Please see the comments. Mainly the `Optional` type hint for a function argument.

File: pykusto/_src/client.py [hunk 2]
```diff
@@ -215,7 +215,7 @@ class Database(_ItemFetcher):
     __name: str
 
     def __init__(
-            self, client: PyKustoClient, name: str, tables: Dict[str, Tuple[BaseColumn]] = None,
+            self, client: PyKustoClient, name: str, tables: Dict[str, Tuple[BaseColumn, ...]] = None,
```
[the remark this answers is not in the document]
Comment (ymost) [reply pykusto/_src/client.py:218]: Yes, but if the default value of the argument is `None`, it is implicitly converted to `Optional[...]`. This is accepted by both the static and runtime type checker, and is more succinct.

File: pykusto/_src/expressions.py [hunk 3]
```diff
@@ -909,22 +933,22 @@ def to_kql(self) -> KQL:
...
 
 class _AssignmentToMultipleColumns(_AssignmentBase):
-    def __init__(self, columns: Union[List['_AnyTypeColumn'], Tuple['_AnyTypeColumn']], expression: ArrayType) -> None:
+    def __init__(self, columns: Iterable['BaseColumn'], expression: ArrayType) -> None:
         super().__init__(KQL(f'({", ".join(c.kql for c in columns)})'), expression)
 
 
 class _AssignmentFromAggregationToColumn(_AssignmentBase):
-    def __init__(self, column: Optional['_AnyTypeColumn'], aggregation: AggregationExpression) -> None:
+    def __init__(self, column: Optional['BaseColumn'], aggregation: AggregationExpression) -> None:
```
[the remark this answers is not in the document]
Comment (ymost) [reply pykusto/_src/expressions.py:951]: It's optional in the sense that the caller can provide `None`, but I do want to force the caller to explicitly provide `None`, to make it less error prone.
The name `Optional` is indeed misleading, it should have been called `Nullable` or something like that.

File: pykusto/_src/functions.py [hunk 4]
```diff
@@ -996,14 +996,14 @@ def to_datetime(expr: StringType) -> _DatetimeExpression:
         return _DatetimeExpression(KQL(f'todatetime({_to_kql(expr)})'))
 
     @staticmethod
-    def to_decimal(expr: NumberType) -> _NumberExpression:
+    def to_decimal(expr: ExpressionType) -> _NumberExpression:
```
[the remark this answers is not in the document]
Comment (ymost) [reply pykusto/_src/functions.py:999]: The `PythonTypes` are supported, since these methods accept literal arguments, for example: `todecimal("123")`.
The argument can also be a general expression, it doesn't have to be as single column, for example: `todecimal(strcat("12", "34"))`

File: pykusto/_src/functions.py [hunk 5]
```diff
@@ -1052,7 +1052,7 @@ def to_lower(expr: StringType) -> _StringExpression:
         return expr.lower()
 
     @staticmethod
-    def to_real(expr: NumberType) -> _NumberExpression:
+    def to_real(expr: ExpressionType) -> _NumberExpression:
```
[the remark this answers is not in the document]
Comment (ymost) [reply pykusto/_src/functions.py:1055]: Same response

File: pykusto/_src/query.py [hunk 6]
```diff
@@ -154,7 +155,7 @@ def project_rename(self, *args: _AssignmentFromColumnToColumn, **kwargs: _AnyTyp
             assignments.append(_AssignmentFromColumnToColumn(_AnyTypeColumn(column_name), column))
         return _ProjectRenameQuery(self, assignments)
 
-    def project_away(self, *columns: StringType) -> '_ProjectAwayQuery':
+    def project_away(self, *columns: Union[BaseColumn, str]) -> '_ProjectAwayQuery':
```
[the remark this answers is not in the document]
Comment (ymost) [reply pykusto/_src/query.py:158]: Good question, I've been wondering about that myself. I think we want to maintain the convenience of the current ways of calling the method: `project_away(t.stringField)` and `project_away("b*")`.
Anyway, if we want to make any improvements, they can wait for another PR.

File: pykusto/_src/query.py [hunk 7]
```diff
@@ -315,7 +316,7 @@ def _extract_assignments(*args: Union[_AssignmentBase, BaseExpression], **kwargs
 
 
 class _ProjectQuery(Query):
-    _columns: List[_AnyTypeColumn]
+    _columns: List[BaseColumn]
```
[the remark this answers is not in the document]
Comment (ymost) [reply pykusto/_src/query.py]: Good point, removed

File: pykusto/_src/query.py [hunk 8]
```diff
@@ -476,18 +477,18 @@ def _compile(self) -> KQL:
...
+        def __init__(self, col: OrderedType, order: Optional[Order], nulls: Optional[Nulls]):
             self.col = col
             self.order = order
             self.nulls = nulls
 
     _query_name: str
     _order_specs: List[OrderSpec]
 
-    def __init__(self, head: Query, query_name: str, col: OrderedType, order: Order, nulls: Nulls):
+    def __init__(self, head: Query, query_name: str, col: OrderedType, order: Optional[Order], nulls: Optional[Nulls]):
```
[the remark this answers is not in the document]
Comment (ymost) [reply pykusto/_src/query.py:490]: See previous responses about the meaning of `Optional` and implicit `Optional`.

File: pykusto/_src/query.py [hunk 9]
```diff
@@ -476,18 +477,18 @@ def _compile(self) -> KQL:
...
         super(_OrderQueryBase, self).__init__(head)
```
[the remark this answers is not in the document]
Comment (ymost) [reply pykusto/_src/query.py]: Done

File: pykusto/_src/query.py [hunk 10]
```diff
@@ -511,15 +512,15 @@ def _compile(self) -> KQL:
 
 
 class _SortQuery(_OrderQueryBase):
-    def __init__(self, head: Query, col: OrderedType, order: Order, nulls: Nulls):
+    def __init__(self, head: Query, col: OrderedType, order: Optional[Order], nulls: Optional[Nulls]):
```
[the remark this answers is not in the document]
Comment (ymost) [reply pykusto/_src/query.py:514]: Same response

File: pykusto/_src/query.py [hunk 11]
```diff
@@ -627,11 +629,11 @@ def _compile(self) -> KQL:
...
     _assignments: Tuple[_AssignmentBase]
-    _bag_expansion: BagExpansion
-    _with_item_index: BaseColumn
-    _limit: int
+    _bag_expansion: Optional[BagExpansion]
+    _with_item_index: Optional[BaseColumn]
+    _limit: Optional[int]
 
-    def __init__(self, head: Query, bag_expansion: BagExpansion, with_item_index: BaseColumn, limit: int, *assignments: _AssignmentBase):
+    def __init__(self, head: Query, bag_expansion: Optional[BagExpansion], with_item_index: Optional[BaseColumn], limit: Optional[int], *assignments: _AssignmentBase):
```
[the remark this answers is not in the document]
Comment (ymost) [reply pykusto/_src/query.py:635]: What would I want the `mv_exapnd` method to omit it?

File: test/test_base.py (hunk 1 above)
Comment (ymost) [reply test/test_base.py:32]: Yes, the database variable in not `Optional`, so we can't provide `None` here

File: pykusto/_src/query.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ofrikleinfeld) [reply pykusto/_src/query.py:158]: I understand the convenience of using `project_away("*")`, but mentioning the binary type `"b*"` is not so convenient.
Can we allow calling `project_away("*")` and convert to binary string ourselves?
As said, it doesn't have to be as part of this PR

File: pykusto/_src/query.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ofrikleinfeld) [reply pykusto/_src/query.py:490]: If we want to enforce the user to explicit supply `None`, then the `then_by` function doesn't need default arguments no?

File: pykusto/_src/query.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ofrikleinfeld) [reply pykusto/_src/query.py:635]: Never mind I confused with something else

File: pykusto/_src/query.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ymost) [reply pykusto/_src/query.py:490]: I don't want to force the user to explicitly provide `None`, that would be inconvenient. But `_OrderQueryBase.__init__` is not called by the user, it's called internally, and this is where I want to force an explicit `None`. Or to look at it differently, there is nothing to gain by making the arguments optional here, since all the arguments need to be provided anyway (look at the call to this method in `_SortQuery.__init__`).

File: pykusto/_src/query.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ymost) [reply pykusto/_src/query.py:158]: The argument here is `"b*"` and not `b"*"`. The result of `project_away("b*")` is projecting away all columns with names that start with `b`.

**Review by ofrikleinfeld — APPROVED** [review]
